xgi/algorithms/clustering.py

```diff
@@ -0,0 +1,233 @@
+from ..linalg import adjacency_matrix
+
+__all__ = [
+    "local_clustering_coefficient",
+    "clustering_coefficient",
+    "two_node_clustering_coefficient",
+]
+
+
+# Properties of metabolic graphs: biological organization or representation artifacts?
+def local_clustering_coefficient(H):
+    """Compute the local clusterin coefficient
+
+    Parameters
+    ----------
+    H : Hypergraph
+        Hypergraph
+
+    Returns
+    -------
+    dict
+        keys are node IDs and values are the
+        clustering coefficients.
+
+    References
+    ----------
+    "Properties of metabolic graphs: biological organization or representation artifacts?"
+    by Wanding Zhou and Luay Nakhleh.
+    https://doi.org/10.1186/1471-2105-12-132
+
+    "Hypergraphs for predicting essential genes using multiprotein complex data"
+    by Florian Klimm, Charlotte M. Deane, and Gesine Reinert.
+    https://doi.org/10.1093/comnet/cnaa028
+
+    Example
+    -------
+    >>> import xgi
+    >>> H = xgi.random_hypergraph(3, [1, 1])
+    >>> cc = xgi.local_clustering_coefficient(H)
+    >>> cc
+    {0: 1.0, 1: 1.0, 2: 1.0}
+    """
+    result = {}
+
+    memberships = H.nodes.memberships()
+    members = H.edges.members()
+
+    for n in H.nodes:
+        ev = list(memberships[n])
+        dv = len(ev)
+        if dv == 0:
+            result[n] = np.NaN
+        elif dv == 1:
+            result[n] = 0
+        else:
+            total_eo = 0
+            # go over all pairs of edges pairwise
+            for e1 in range(dv):
+                edge1 = members[e1]
+                for e2 in range(e1):
+                    edge2 = members[e2]
+                    # set differences for the hyperedges
+                    D1 = set(edge1) - set(edge2)
+                    D2 = set(edge2) - set(edge1)
+                    # if edges are the same by definition the extra overlap is zero
+                    if len(D1.union(D2)) == 0:
+                        eo = 0
+                    else:
+                        # otherwise we have to look at their neighbours
+                        # the neighbours of D1 and D2, respectively.
+                        neighD1 = {i for d in D1 for i in H.nodes.neighbors(d)}
+                        neighD2 = {i for d in D2 for i in H.nodes.neighbors(d)}
+                        # compute the extra overlap [len() is used for cardinality of edges]
+                        eo = (
+                            len(neighD1.intersection(D2))
+                            + len(neighD2.intersection(D1))
+                        ) / len(
+                            D1.union(D2)
+                        )  # add it up
+                    # add it up
+                    total_eo = total_eo + eo
+
+            # include normalisation by degree k*(k-1)/2
+            result[n] = 2 * total_eo / (dv * (dv - 1))
+    return result
+
+
+def clustering_coefficient(H):
+    """Return the clustering coefficients for
+    each node in a Hypergraph.
+
```

Comment:
Here the definition is `cc[i] = A^3[i,i] / (k[i] * (k[i] - 1))` correct?

Reply:
Almost. Instead of k, it's the number of neighbors.